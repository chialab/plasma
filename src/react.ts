import { mkdir, writeFile } from 'node:fs/promises';
import { dirname, join } from 'node:path';
import type { Entry } from './walker';

export interface ReactTransformOptions {
    /**
     * The output directory to write the converted components to.
     */
    outdir: string;
}

function generateReactComponent(entry: Entry) {
    const { packageJson, definition, declaration } = entry;
    let props = '';
    if (declaration.members) {
        for (const member of declaration.members) {
            if (member.kind !== 'field') {
                continue;
            }
            if (member.privacy && member.privacy !== 'public') {
                continue;
            }
            props += `${member.name}, `;
        }
    }

    return `import React, { useRef, useEffect } from 'react';
import '${packageJson.name}';

export const ${declaration.name} = ({ children, ${props}...props }) => {
    const ref = useRef();

    useEffect(() => {
        Object.assign(ref.current, { ${props} });
    });

    return React.createElement(${definition.extend ?? definition.name}, {
        ref,
        ...props,
    }, ...children);
`;
}

function generateReactTypings(entry: Entry) {
    return '';
}

export async function transformReact(entry: Entry, options: ReactTransformOptions) {
    const { declaration } = entry;
    const outFile = join(options.outdir, `${declaration.name}.js`);
    const declFile = `${outFile}.d.ts`;

    await mkdir(dirname(outFile), {
        recursive: true,
    });
    await Promise.all([
        writeFile(outFile, generateReactComponent(entry)),
        writeFile(declFile, generateReactTypings(entry)),
    ]);

    return outFile;
}
