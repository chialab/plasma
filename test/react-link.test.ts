import React from 'react';
import { createRoot } from 'react-dom/client';
import { afterEach, beforeEach, describe, expect, test, vi } from 'vitest';
import { TestLink } from './src/react/TestLink';

let host: HTMLElement;

const awaitReactRender = () => new Promise((resolve) => setTimeout(resolve, 100));

describe('Element', () => {
    beforeEach(() => {
        host = document.createElement('div');
        document.body.appendChild(host);
    });

    afterEach(() => {
        host.remove();
    });

    test('mount component', async () => {
        const root = createRoot(host);
        const onStringChange = vi.fn();
        const onClick = vi.fn((event) => event.preventDefault());
        root.render(
            React.createElement(TestLink, {
                'stringProp': 'test',
                'booleanProp': true,
                'numericProp': 1,
                'objectProp': { test: true },
                'data-attr': 'test',
            })
        );
        await awaitReactRender();
        const node = host.querySelector('a') as HTMLElement;
        expect(host.innerHTML).toMatchSnapshot();
        expect(onStringChange).not.toHaveBeenCalled();
        expect(onClick).not.toHaveBeenCalled();
        root.render(
            React.createElement(TestLink, {
                'onStringchange': onStringChange,
                onClick,
                'stringProp': 'changed',
                'booleanProp': true,
                'numericProp': 1,
                'objectProp': { test: true },
                'data-attr': 'test',
            })
        );
        await awaitReactRender();
        expect(onStringChange).toHaveBeenCalled();
        node.click();
        expect(onClick).toHaveBeenCalled();
    });
});
