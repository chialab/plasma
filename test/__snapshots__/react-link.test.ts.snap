// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Element > mount component 1`] = `"<a is=\\"test-link\\" data-attr=\\"test\\" :scope=\\"test-link\\" :defined=\\"\\" stringprop=\\"test\\" booleanprop=\\"\\" numericprop=\\"1\\"><span class=\\"text\\">test</span><span class=\\"boolean\\">true</span><span class=\\"number\\">1</span><span class=\\"object\\">{\\"test\\":true}</span></a>"`;
